### PR TITLE
Allow rotation of camera image

### PR DESCRIPTION
**Pull request by max-scopp**
Really needed that feature.

I can also make it so the black bars are avoided, but this would require more effort in resizing the container.

Requires further testing on IP Cameras and HTTP pages.
HTTP pages might cause weird behavior due to the `height` setting.
IP cameras might cause weird behavior due to visible controls of the HTML5 video tag.

Tested:
- [x] MJPEG Adaptive
- [x] MJPEG Stream
- [ ] IP Camera
- [ ] HTTP page

Resolves https://github.com/fluidd-core/fluidd/issues/33

image

### CODE REVIEW

**Comment by grrminator** [comment]
how does one put this on fluiddpi?

**Comment by max-scopp** [comment]
Either run a

**Comment by max-scopp** [comment]
It's not a good (non-permanent) solution, but you can clone my fork on your computer, build the frontend and copy the resulting contents of the created `dist/` folder onto your pi. Klipper is usually installed under `~/fluidd`. However, I'm not sure how the updater would handle this.

Alternatively, you can create a web server somewhere and access it from there. Simply go through the install process again and point it to the pi's ip/host.

**Comment by point-source** [comment]
What is this still waiting on for a proper merge? Review?

File: src/components/settings/cameras/CameraConfigDialog.vue [hunk 1]
```diff
@@ -58,6 +58,40 @@
...
+              {
+                text: $t('app.setting.camera_rotate_options.180'),
+                value: '180',
+              },
+              {
+                text: $t('app.setting.camera_rotate_options.270'),
+                value: '270',
+              },
+              {
+                text: $t('app.setting.camera_rotate_options.360'),
```
Comment (matmen) [inline src/components/settings/cameras/CameraConfigDialog.vue:82]: I don't really see a need for a 360° rotation

File: src/components/widgets/camera/CameraItem.vue [hunk 2]
```diff
@@ -77,16 +70,33 @@ export default class CameraItem extends Vue {
...
-    transforms += (config && config.flipX) ? ' scaleX(-1)' : ''
-    transforms += (config && config.flipY) ? ' scaleY(-1)' : ''
-    // transforms += ` rotate(${config.rotate}deg)`
-    return (transforms.trimLeft().length) ? { transform: transforms.trimLeft() } : {}
+
+    if (!config.rotate) {
+      transforms += config && config.flipX ? ' scaleX(-1)' : ''
+      transforms += config && config.flipY ? ' scaleY(-1)' : ''
+    } else {
+      let scaling = element.clientHeight / element.clientWidth
```
Comment (matmen) [inline src/components/widgets/camera/CameraItem.vue:87]: the scaling works fine for 90° and 270° rotation, though i am not sure about the black bars (they are probably not easily removable because of the footer text tho; and maybe even fine considering each camera view is meant to split the card evenly):
90 deg rotation

for 180° (and 360°) the scaling is quite off:
180 deg rotation

File: src/locales/de.yaml [hunk 3]
```diff
@@ -262,9 +262,16 @@ app:
       mjpegadaptive: MJPEG Adaptive
       mjpegstream: MJPEG Stream
       video: IP Camera
+    camera_rotate_options:
+      none: Keine
+      90: 90°
+      180: 180°
+      270: 270°
+      360: 360°
```
Comment (matmen) [inline src/locales/de.yaml:270]: see first comment

File: src/locales/en.yaml [hunk 4]
```diff
@@ -335,9 +335,16 @@ app:
       mjpegstream: MJPEG Stream
       video: IP Camera
       iframe: HTTP page
+    camera_rotate_options:
+      none: None
+      90: 90°
+      180: 180°
+      270: 270°
+      360: 360°
```
Comment (matmen) [inline src/locales/en.yaml:343]: see first comment

File: src/locales/en.yaml [hunk 5]
```diff
@@ -335,9 +335,16 @@ app:
...
     label:
       all_off: All off
       all_on: All on
+      camera_rotate_by: Rotate By
```
Comment (matmen) [inline src/locales/en.yaml:347]: probably lower case "by"?

File: src/store/cameras/types.ts [hunk 6]
```diff
@@ -12,5 +12,6 @@ export interface CameraConfig {
   fpstarget?: number;
   flipX: boolean;
   flipY: boolean;
+  rotate?: '90' | '180' | '270' | '360';
```
Comment (matmen) [inline src/store/cameras/types.ts:15]: see first comment

File: src/locales/de.yaml [hunk 7]
```diff
@@ -262,9 +262,16 @@ app:
       mjpegadaptive: MJPEG Adaptive
       mjpegstream: MJPEG Stream
       video: IP Camera
+    camera_rotate_options:
+      none: Keine
```
Comment (matmen) [inline src/locales/de.yaml:266]: "Drehen um: Keine" seems a bit off, maybe replace this with "Keine Drehung" to be more verbose?

**Review by matmen — CHANGES_REQUESTED** [review]

File: src/components/widgets/camera/CameraItem.vue (hunk 2 above)
Comment (max-scopp) [reply src/components/widgets/camera/CameraItem.vue:87]: Yes, there are some questionable cases that would require scaling the parent elements, so I didn't intend to change too much. Maybe I'll try it for the next iteration.

My camera is currently not usable on my printer, but I'll check this presumably Monday/Tuesday evening. I also tried it with my 1080p camera, which worked on a maximized browser window. Can you leave your browser resolution and just in case the camera resolution? I don't see a reason it would be different as long as the aspect ratio stays the same, but oh well.

Comment (matmen) [reply src/components/widgets/camera/CameraItem.vue:87]: I'm using Chrome on Win10 on a 1080p monitor with a 1080p camera. I assume scaling at 180° can be omitted completely, because it will just result in the same canvas size.

**Review by matmen — DISMISSED** [review]
LGTM

image

**Comment by matmen** [comment]
One small issue, actually: When collapsing the camera view card, the animation frame handler throws `Uncaught TypeError: Cannot read properties of undefined (reading 'style')`. Probably just needs a `if(!image) return;`, though.

Also I'm not quite sure about the recursive call of animate, as it pushes to the stack with every frame. Not sure if Vue is doing some magic in the background to keep the stack from overflowing (need some time to test this, my browsers max stack size is 13953, so 15 minutes at 15fps for it to overflow. I'll comment on here when I know more).

**Comment by matmen** [comment]
Doesn't seem like it overflows. Just the TypeError then. Nice work!

**Comment by max-scopp** [comment]
Yikes! Totally forgot about collapsing. Good eye on the QA though!

A simple return would have worked, but the current solution is more appropriate since the error only occurs when trying to transform a frame where the component was already being unmounted/destroyed. But, just in case, I added the missing `undefined` to the refs' type.

**Review by matmen — APPROVED** [review]
LGTM

**Review by matmen — APPROVED** [review]

**Comment by matmen** [comment]
docs: see https://github.com/fluidd-core/fluidd/pull/500#issuecomment-1048247653
